core/src/memory.rs

```diff
@@ -367,49 +543,80 @@ impl AllocationsContext {
         &mut self,
         pages: WasmPageNumber,
         mem: &mut impl Memory,
-    ) -> Result<WasmPageNumber, Error> {
-        let mut previous = None;
-        let mut current = None;
-
-        let mut at = None;
-
-        let last_static_page = (self.static_pages.0 != 0).then(|| self.static_pages - 1.into());
-        for page in last_static_page.iter().chain(self.allocations.iter()) {
-            if current.is_some() {
-                previous = current;
+    ) -> Result<AllocInfo, Error> {
+        let mem_size = mem.size();
+        let mut previous = self.static_pages;
+        let mut start = None;
+        for &page in self.allocations.iter().chain(core::iter::once(&mem_size)) {
+            if page
+                .sub(previous)
+                .map_err(|_| Error::IncorrectAllocationsSetOrMemSize)?
+                >= pages
+            {
+                start = Some(previous);
+                break;
             }
 
-            current = Some(page);
-
-            if let Some(&previous) = previous {
-                if (*page).saturating_sub(previous) > pages {
-                    at = Some((previous).saturating_add(1.into()));
-                    break;
-                }
-            }
+            previous = page.inc().map_err(|_| Error::OutOfBounds)?;
         }
 
-        let at = at
-            .or_else(|| current.map(|v| (*v).saturating_add(1.into())))
-            .unwrap_or(self.static_pages);
+        let (start, not_grown) = if let Some(start) = start {
+            (start, pages)
+        } else {
+            // If we cannot find interval between already allocated pages, then try to alloc new pages.
+
+            // Panic is safe, because we check, that last allocated page can be incremented in loop above.
+            let start = self
+                .allocations
+                .last()
+                .map(|last| last.inc().unwrap_or_else(|err| {
+                    unreachable!("Cannot increment last allocation: {}, but we checked in loop above that it can be done", err)
+                }))
+                .unwrap_or(self.static_pages);
+            let end = start.add(pages).map_err(|_| Error::OutOfBounds)?;
+            if end > self.max_pages {
```

Comment:
```suggestion
            if end >= self.max_pages {
```
end is ptr, max_pages is count, isn't it?

Reply:
No, because `end` is excluded page number.  allocated_pages == [start, end)